### ISSUE TITLE
TestPlan: Read new debugger docs

### ISSUE THREAD

**Issue by rchiodo** [microsoft/vscode-jupyter#10917]
Validating new wiki docs on Debugger

Refs: https://github.com/microsoft/vscode-jupyter/wiki/How-debuggers-work

- [ ] anyOS @rebornix
- [ ] anyOS @amunger 

Complexity: 1

Authors: @rchiodo

---

Just read the documents and either fix the wiki, ask questions here, or add issues about anything that doesn't make any sense.

**Comment by amunger** [comment]
https://github.com/microsoft/vscode-jupyter/wiki/How-interactive-debugging-works-(old-way) references an "IP indicator", not sure what that is - Instruction pointer?

**Comment by rchiodo** [comment]
Yes it comes from `eip`, the register that has the instruction pointer in it. Should I reword it to 'instruction pointer'? I wonder if the VS code docs talk about it.

**Comment by rchiodo** [comment]
I added a link to wikipedia about it but not sure that makes it clearer. I cant' find any reference to it in the VS code docs other than an image.

**Comment by amunger** [comment]
from https://github.com/microsoft/vscode-jupyter/wiki/How-notebook-debug-cell-works
> [dumpCell](https://jupyter-client.readthedocs.io/en/stable/messaging.html#additions-to-the-dap). This takes the contents of a cell and generates a temporary file for it

Does the `dumpCell` call actually generate the file, or just return the location of the already generated file? Would it just use the notebook file without the call to dump cell?

**Comment by amunger** [comment]
It sounds like we manually inject debugby into IPykernel with IW debugging, 

> The first step is to inject debugpy into kernel

But it does it on its own for notebook debugging:

> Now you might wonder did IPython implement its own debugger?
> No, it actually loads debugpy into itself and forwards DAP requests to debugpy.

**Comment by amunger** [comment]
copy pasta, or does it really translate filepaths twice? I don't see an explanation for that
![image](https://user-images.githubusercontent.com/2019016/181063290-df2e41a3-e23e-4bfc-896c-dc4d12170241.png)

**Comment by rebornix** [comment]
@rchiodo thanks for the well documented wiki, it gives good overview of how debugger works with right amount of technical details. 

I only have two quick questions which might not be covered in the wiki:

* Inject debugpy https://github.com/microsoft/vscode-jupyter/wiki/How-interactive-debugging-works-(old-way)#inject-debugpy
  * How do we inject `debugpy`? Do we ship `debugpy` with the extension or it's installed from `pip`?
  * Is there any difference between local raw kernel and remote kernels? 
  * Do we inject `debugpy` for all `ipykernel` versions? Is it still required when Jupyter kernel comes with builtin DAP support?
* Remapping file names https://github.com/microsoft/vscode-jupyter/wiki/How-interactive-debugging-works-(old-way)#remapping-file-names
  * When we remap the file name, do we also map the `line:col` positions? Do we remap `line:col` in https://github.com/microsoft/vscode-jupyter/wiki/How-notebook-debug-cell-works#translate-file-paths?

**Comment by rchiodo** [comment]
> from https://github.com/microsoft/vscode-jupyter/wiki/How-notebook-debug-cell-works
> 
> > [dumpCell](https://jupyter-client.readthedocs.io/en/stable/messaging.html#additions-to-the-dap). This takes the contents of a cell and generates a temporary file for it
> 
> Does the `dumpCell` call actually generate the file, or just return the location of the already generated file? Would it just use the notebook file without the call to dump cell?

I'm not sure what you're asking. Dump cell generates the file. The kernel generates the file. If we don't do that, then we just end up with a random file name and we cannot predict that file name, so the 'TranslateFilePaths' wouldn't work.

**Comment by rchiodo** [comment]
And translatefilepaths is called on every request. It has to because the request contains invalid file paths. I added a blurb about that.

**Comment by rchiodo** [comment]
> It sounds like we manually inject debugby into IPykernel with IW debugging,

> The first step is to inject debugpy into kernel

> But it does it on its own for notebook debugging:

> Now you might wonder did IPython implement its own debugger?
> No, it actually loads debugpy into itself and forwards DAP requests to debugpy.

> Is that actually the case?

Yes. That's actually the case. When you send a 'requestDebug' request, it will inject debugpy.

**Comment by rchiodo** [comment]
To answer @rebornix

> Inject debugpy https://github.com/microsoft/vscode-jupyter/wiki/How-interactive-debugging-works-(old-way)#inject-debugpy
How do we inject debugpy? Do we ship debugpy with the extension or it's installed from pip?

I added a blurb here: 

https://github.com/microsoft/vscode-jupyter/wiki/How-interactive-debugging-works-(old-way)#where-does-debugpy-come-from

> Is there any difference between local raw kernel and remote kernels?
Remote is not supported. I added a small bit of text here: https://github.com/microsoft/vscode-jupyter/wiki/How-interactive-debugging-works-(old-way)#what-about-remote

> Do we inject debugpy for all ipykernel versions? Is it still required when Jupyter kernel comes with builtin DAP support?
Yes for now. Don was working on using the built in DAP support in ipykernel 6, but it's flakey at the moment.

**Comment by rchiodo** [comment]
> Remapping file names https://github.com/microsoft/vscode-jupyter/wiki/How-interactive-debugging-works-(old-way)#remapping-file-names
When we remap the file name, do we also map the line:col positions? Do we remap line:col in https://github.com/microsoft/vscode-jupyter/wiki/How-notebook-debug-cell-works#translate-file-paths?

Not for notebook debugging. It shouldn't be necessary as the temp file is the contents of the cell (each cell is a unique URI).

For the new IW ipykernel 6 debugging (behind a flag at the moment), it does [translate line numbers](https://github.com/microsoft/vscode-jupyter/blob/c68a25624fb3f46656795ef8a4d66a2db8e368ca/src/interactive-window/debugger/jupyter/kernelDebugAdapter.ts#L120) (columns shouldn't be necessary AFAIK).
